### ISSUE TITLE
[Question][Hw3] About Division in Z3

### ISSUE THREAD

**Issue by GetRusty** [prosyslab-classroom/cs492-program-reasoning#167]
Name: Ingyu Baeg, 백인규

Dear All,
For HW3 I tried to put some custom examples to maintain code coverage, but strangely Z3 rejected my interpretation of the smt2 file of the formula. My custom code consisted of some long sequences, plus assertion of some division at the end.
Apparently Z3 cannot understand the "div" operator in my .smt2 file, and as shown in the pictures below I believe that I did not misbehave in interpreting division into the formula. Also I could not find appearances of "div" in any other locations of the formula.

I changed my c code simpler but the same problem remains; so I just pushed it to github and gradescope and it got full score. Maybe Z3 has some internal mistakes for reading the "div" operator? Curious about this.

![image](https://github.com/prosyslab-classroom/cs492-program-reasoning/assets/42976401/6e8033fc-973c-4c50-9c7e-39909af025a2)
![image](https://github.com/prosyslab-classroom/cs492-program-reasoning/assets/42976401/42227adf-73a4-4a74-b466-da544f9cd4cd)
![image](https://github.com/prosyslab-classroom/cs492-program-reasoning/assets/42976401/03ffc57f-e8eb-48f2-a053-1194aba8c7f6)
![image](https://github.com/prosyslab-classroom/cs492-program-reasoning/assets/42976401/84b52d21-b118-4d52-8c95-4a4e03dd41fd)

**Comment by goodtaeeun** [comment]
Hi, just for clarification, is this .smt2 file hand-written?

**Comment by GetRusty** [comment]
> Hi, just for clarification, is this .smt2 file hand-written? Furthermore, can you share the full .smt2 file in this example?

No, It is generated by executing verifier on my custom example testcase. 

#include "verifier.h"

int main()
{
    int x = input();
    int y = input();
    int w = y - x;
    assume(x > 1);
    assume(y > 49);
    assume(x < 5);
    int f = x / w;
    int q = 5;
    assert(f <= q);
    return 0;
}
![image](https://github.com/prosyslab-classroom/cs492-program-reasoning/assets/42976401/c1e99dcc-ad93-40b7-835f-de0edf99f937)

I get an error like this.
the below .txt file is an .smt2 file of the above c code.



[debug.txt](https://github.com/prosyslab-classroom/cs492-program-reasoning/files/13159118/debug.txt)
[formula.txt](https://github.com/prosyslab-classroom/cs492-program-reasoning/files/13159123/formula.txt)

**Comment by goodtaeeun** [comment]
Hi, I did some searching and found this issue (https://github.com/Z3Prover/z3/issues/3962).
According to the developers, the Horn clause solvers can't handle `div` function when the second argument is a variable.

**Comment by GetRusty** [comment]
> Hi, I did some searching and found this issue ([Z3Prover/z3#3962](https://github.com/Z3Prover/z3/issues/3962)). According to the developers, the Horn clause solvers can't handle `div` function when the second argument is a variable.

Oh! very interesting. Then it's not really a great idea to add custom cases with variable divisions. Thanks for the searching sir.
